Add a node whitelist filter

alexmavr opened the pull request:
This PR adds support for a new scheduling filter that selects candidates based on a provided list of nodes, effectively implementing a "whitelist". This is meant to cover use cases where setting engine labels may not be practical.

The usage is currently implemented as `-e whitelist:node==node-1-name|node-2-name|node-3-name`, and I am definitely open to suggestions for a better UX.

A similar approach that is available today is to use a series of `constraint:name!=<name>` constraints to implement a "blacklist" of nodes. However, that method is not equivalent to a whitelist filter during rescheduling.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "whitelist-filter" git@github.com:alexmavr/swarm.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

nishanttotla commented:
@alexmavr can you please rebase this PR, it would be nice to have it on top of #2492.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "whitelist-filter" git@github.com:alexmavr/swarm.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

nishanttotla commented on scheduler/filter/filter.go:
```diff
@@ -33,6 +33,7 @@ func init() {
 		&DependencyFilter{},
 		&AffinityFilter{},
+		&WhitelistFilter{},
 		&ConstraintFilter{},
```
Are you sure no more changes are needed in `filter.go`? I'm looking more but notice that filters are applied serially, since each filter only serves to further reduce the number of available nodes. The whitelist filter has a fundamentally opposite purpose, so it should be applied last, and the final set should be a union of the result of old filters with whitelisted nodes.

alexmavr replied:
I don't see why ordering would matter, since they're all individual filters. Whitelist is not meant to circumvent any other filters of the request, but it's meant to add an additional one, in case a user wants to specify a specific subset of nodes where the rest of the filters would be applied.

I'll go ahead and put it last either way though :)

nishanttotla replied:
Ah! My bad actually, I commented before reading through the rest of the code. I misunderstood what this filter was supposed to do (whitelist to me implied that it's a whitelist of nodes indicating that regardless of the outcome of other filters, a whitelisted node should always be allowed to run the task).

I now understand, and sorry for the confusion.

nishanttotla commented:
The whitelist functionality is definitely new and something that isn't supported in Swarm right now. One concern I have about it is that it puts more responsibility on the user, to make sure that (other) hard constraint requirements aren't accidentally overridden by the whitelist filter. Essentially, it comes with a warning to use at your own risk.

cc @dongluochen

nishanttotla commented:
@alexmavr in addition, do you mind rebasing one more time? Several tests were fixed recently.

alexmavr commented:
Rebased and addressed line comment (moved whitelist filter last).

I still don't understand why you say this is a different case for a filter, the use case is to have all types of filters apply, but ignore any candidate nodes that are not part of the whitelisted nodes.

dongluochen commented on scheduler/filter/whitelist.go:
```diff
@@ -0,0 +1,71 @@
+		return nil, err
+	}
+
+	for _, whitelist := range whitelists {
```
what's the use case of multiple whitelists for a `docker run` command?

dongluochen replied:
What does it mean when running 
```
docker run --label 'com.docker.swarm.whitelists=["node==node1,node2,node3"]' --label 'com.docker.swarm.whitelists=["node==node2,node3,node4"]'
```

dongluochen replied:
At a second thought, this way is more consistent with other filters like `constraints`. Let's keep it as it is.

dongluochen commented:
Once we merge this change, please update Swarm doc to reflect the change.

dongluochen commented:
LGTM

Tests failure looks unrelated. Should investigate separately.

dongluochen commented:
ping @nishanttotla @allencloud.

nishanttotla commented:
Restarted the CI to run tests one more time.

Code LGTM.

nishanttotla commented:
LGTM

@alexmavr please submit a PR to update docs.

@mstanleyjones
